### ISSUE TITLE
Feature: Syntax highlighting for custom rules

### ISSUE THREAD

**Issue by alexanderadam** [xvello/letsblockit#68]
### Describe the bug

It would be great if syntax highlighting for custom rules (see [[this JS library](https://github.com/RedDragonWebDesign/adblock-filter-analyzer)](https://github.com/RedDragonWebDesign/adblock-filter-analyzer) for instance) could be added.

### Your browser and device

-

### Reproduction steps

```bash
This is not a bug but there's no feature category available. Only bug, filter adding and filter updating.
```

**Comment by xvello** [comment]
That would be a great UX improvement indeed. There are other options I'd love to investigate:
- Reusing [uBlockO's custom CodeMirror config](https://github.com/gorhill/uBlock/tree/master/src/js/codemirror) for consistency with the plugin, although it's not the best syntax highlighting I worked with
- Maintain a custom syntax for the lightweight [Prism syntax highlighting engine](https://github.com/PrismJS/prism/). It has a work-in-progress [edition support](https://github.com/PrismJS/live)

Help is very welcome on this one!

**Comment by alexanderadam** [comment]
Apart from Syntax highlighting itself it would be nice to have warnings if someone tries to add something with invalid syntax.
Would think that this should be handled in the same issue or would that be a separate one?

**Comment by xvello** [comment]
I think it's better to re-scope this issue as an "advanced editor" one. If we want to go as far as detecting syntax errors, we'll definitely need to reuse common work.

You previously linked to https://github.com/RedDragonWebDesign/adblock-filter-analyzer, would you have some cycles available to reach out to the author and discuss publishing their code under a permissive (MIT / Apache2) license (their project is currently "all rights reserved") and investigate how to integrate it in the letsblock.it UX?

The other option is using uBO's upstream parser, but it being GPLv3 licensed will make it a bit harder (it explicitly chose the Apach2 license for [clause 5](https://github.com/xvello/letsblockit/blob/3e9bb0abcfae894ee685e355c0142fcee8b7b505/LICENSE.txt#L131-L137), which makes project governance easier without a CLA).

**Comment by xvello** [comment]
I have been looking at CodeMirror 6's new syntax definition system, and it's pretty impressive how small their [css language definition](https://github.com/codemirror/lang-css) is. Seeing how many extensions this editor comes with, I'm very interested in trying to POC an adblock language definition for CM6 (that we could maintain in a separate repo for reuse, and import in LBI). Is that something you'd be interested in doing, or shall I give it a go sometime?

If we manage to do this, the editor would be a separate [parcel asset target](https://github.com/letsblockit/letsblockit/blob/a5a1b03d7956682878f4a5bffb0913bd616067a0/src/assets/package.json#L5) that we would load on the custom-rules page to replace the textarea. After it has baked for a while, we could also use it for the rendered output of other filters.

https://codemirror.net/examples/lang-package/
https://lezer.codemirror.net/docs/guide/#writing-a-grammar
https://marijnhaverbeke.nl/blog/lezer.html

**Comment by alexanderadam** [comment]
I surely have no idea what the best way of doing this would be but I know that it is something I couldn't do right now.

> * Reusing [uBlockO's custom CodeMirror config](https://github.com/gorhill/uBlock/tree/master/src/js/codemirror) for consistency with the plugin, although it's not the best syntax highlighting I worked with

Maybe it'd be better than not having anything at all?
What's your feeling about this?

**Comment by xvello** [comment]
Indeed, this could be a good first version, but probably not the best base to iterate on top of (it uses the extension's rule parsing lib for code reuse, but that's probably not the best tokenization one can do). The HTML for the "My filters" page is https://github.com/gorhill/uBlock/blob/master/src/1p-filters.html
